### PR TITLE
Document JSON output (#31)

### PR DESCRIPTION
**Pull request by willkg**
Couple of thoughts:

1. this is probably missing things
2. this doesn't differentiate between what Socorro uses vs. what's vestigial
3. this doesn't take into account changes we're going to make for #30--I figure we can update the README accordingly with that PR

### CODE REVIEW

**Comment by willkg** [comment]
@Gankra I figure while it's nice to have the schema in the rust-minidump repository issues, it'd be best if we "officially" documented it in this repo.

I went through a bunch of crash reports and the stackwalker.cc code and some other things and fleshed out the schema you threw together. My gut feeling is that it's probably missing things, still, but we can hone it in this PR and then land it and iterate on that.

What do you think?

File: README.rst [hunk 1]
```diff
@@ -112,6 +112,113 @@ By default both minidump-stackwalk and breakpad are built with PGO and an
...
+      }
+    ],
+    
+    // this is a repeat of the crashing thread in the threads list, but the
+    // number of frames is truncated to 10
+    "crashing_thread": {
+      "threads_index": <int>,                  // index in threads for the crashing thread
+      "total_frames": <int>,                   // total frames listed here -- max 10
+      "frames": [
+        // for i in range(min(10, frame_count))
```
Comment (Gankra) [inline README.rst]: I thought we removed this truncation because it was confusing?

**Review by Gankra — COMMENTED** [review]
This is great, thanks!

**Review by Gankra — APPROVED** [review]

File: README.rst (hunk 1 above)
Comment (willkg) [reply README.rst]: I forgot about that! Good catch!

https://github.com/mozilla-services/minidump-stackwalk/commit/5896847652369013057cf7d1733400f616b2f8aa

**Comment by willkg** [comment]
I fixed the thing @Gankra noticed and also added a bunch of other stuff I saw while reading through stackwalker.cc including the new `mac_crash_info` bits, the `lsb_release` bits, and a bunch of other bits.

File: README.rst [hunk 2]
```diff
@@ -112,6 +112,146 @@ By default both minidump-stackwalk and breakpad are built with PGO and an
...
+        "code_id": <string>,
+        "version": <string>,
+        "loaded_symbols": true,              // optional, if mdsw looked for the file and it does exist
+        "missing_symbols": true,             // optional, if mdsw looked for the file and it doesn't exist
+        "corrupt_symbols": true,             // optional, if mdsw found a file that has parse errors
+        "symbol_disk_cache_hit": <bool>,     // optional, whether or not the SYM file was fetched from disk cache
+        "symbols_fetch_time": <float>,       // optional, time in ms it took to fetch symbol file from url; omitted
+                                             // if the symbol file was in disk cache
+        "symbol_url": <string>,              // optional, url of symbol file
+        "cert_subject": ?                    // optional
```
Comment (willkg) [inline README.rst]: I couldn't figure out what this was and I haven't seen examples in crash reports processed by Socorro.

@gabrielesvelto , @Gankra : Any ideas what this is?

Comment (gabrielesvelto) [reply README.rst]: The `cert_subject` field contains the entity that signed a given module using Microsoft's Authenticode system. For our modules for example this is going to be "Mozilla Corporation", Microsoft own modules will be signed by "Microsoft Windows" or "Microsoft Corporation" and so on. The minidump-analyzer [retrieves this information](https://searchfox.org/mozilla-central/rev/98a9257ca2847fad9a19631ac76199474516b31e/toolkit/crashreporter/minidump-analyzer/minidump-analyzer.cpp#217-241) when it processes a crash and stuffs it in an annotation called [ModuleSignatureInfo](https://searchfox.org/mozilla-central/rev/98a9257ca2847fad9a19631ac76199474516b31e/toolkit/crashreporter/minidump-analyzer/minidump-analyzer.cpp#493).

Socorro's stackwalker uses this info to [populate that field](https://github.com/mozilla-services/minidump-stackwalk/blob/main/minidump-stackwalk/stackwalker.cc#L706-L709).

**Comment by Gankra** [comment]
Note gabriel found several errors here (I copy-pasted your schema): https://github.com/luser/rust-minidump/pull/170#issuecomment-843465955

**Comment by gabrielesvelto** [comment]
FYI I compared the output directly with what we get in Socorro. So mine was one of those "the implementation is the spec" kind of things.

**Comment by willkg** [comment]
I double-checked and those are all errors. I'll fix and then land this. Thank you both!

**Pull request merged**